### DISCUSSION TITLE
Headless Exception on Code Server with Ipad Pro and Raspberry Pi

### DISCUSSION THREAD

**Discussion by nigelpurvis** [coder/code-server#5252] [Q&A] [answered]
I am using Code Server on my iPad with a Raspberry Pi and I am running a Java Project. I downloaded JDK onto my Raspberry Pi and when I run my program it says: "No X11 DISPLAY variable was set, but this program performed an operation which requires it."

I am assuming it has something to do with me being on iPad.
Also, the program I am running requires User Inputs from the keyboard, and I am assuming this also may have something to do with it.

I would appreciate the help!

**Comment by lg-nineoneninethreenine** [comment]
Have you tried enabling VNC and using that? Could help

**Reply by nigelpurvis** [reply]
Oh hmmm maybe, except I'm not sure how I would do that on my iPad haha.

**Reply by lg-nineoneninethreenine** [reply]
Well, VNC Viewer or better yet, noVNC!

**Comment by jsjoeio** [comment]
How are you running code-server and accessing via your iPad?

**Reply by nigelpurvis** [reply]
Plug the Raspberry Pi with slightly modified OS into my iPad, SSH Into Pi with Blink Shell, and Run Code-Server. I followed a couple of videos from TechCraft, super super helpful and super interesting! 

Link: https://www.youtube.com/watch?v=11YfaGi0Fpk

Also the OS I downloaded onto the Raspberry Pi is a little different, I'm not exactly sure how but TechCraft said something about how it helps connect the Pi to my iPad. 
Link: https://github.com/techcraftco/rpi-usb-gadget/releases

**Comment by nigelpurvis** [comment, marked as answer]
Nvm, I fixed it, the program I was running was accidentally calling a different function with a JavaSwing GUI. The GUI was probably the reason for throwing the error since the iPad can't run the GUI for some reason. But now it works and I can run VSCode on my iPad!

**Comment by teja-self-taught-cs** [comment]
I have connected pi to my ipad, now trying to install code server But I'm getting this error !
Please 🙏 let me know I'm stuck..
How to solve this.....
![IMG_20221102_131112](https://user-images.githubusercontent.com/104780299/199440686-eaa858c9-845b-4a25-aad2-1587a07e6fd8.jpg)

**Reply by jsjoeio** [reply]
the solution is in the last line "use npm to install code-server" - `npm -g install code-server --unsafe-perm`
